// @skinternal/skconnectorsdk/config/constants.js
/**
 * Created by hiteshkalra on 5/21/18.
 */

exports.customersTable = "Customers";
exports.indexCustomers_email = "indexCustomers_email";
exports.appsTable = "Apps";
exports.indexApps_apiKeys_prod = 'indexApps_apiKeys_prod';
exports.indexApps_companyId = 'indexApps_companyId';
exports.appUsersTable = 'Users'
exports.indexAppUsers_username_companyId_appId = 'indexAppUsers_username_companyId_appId';
exports.deviceUsersTable = 'Devices';
exports.eventsTable = 'Events';
exports.configTable = 'Config';
exports.mdsTable = 'Mds';
exports.connectorsTable = 'Connectors';
exports.connectionsTable = 'Connections';
exports.companyTable = 'Company';
exports.flowsTable = 'Flows';
exports.constructsTable = 'Constructs';
exports.flowVersionsTable = 'FlowVersions'

exports.redisATPrefix = "skauth-";
exports.redisATMfaRequiredPrefix = 'skauthmfarequired-';
exports.redisDeviceATPrefix = "skdeviceauth-";
exports.redisChallengePrefix = "skchallenge-";
exports.redisTotpCredentialPrefix = "sktotp-";
exports.redisOtpCredentialPrefix = "skOtp-";

exports.redisStateTokenPrefix = 'skstatetoken-';
exports.redisCustomerVerificationCodePrefix = 'skcustomerverificationcode-';

exports.redisOpenIdAuthPageState = 'skopenidauthpagestate-';
exports.redisOpenIdCode = 'skopenidcode-';
exports.redisOpenIdRefreshToken = 'skopenidrefreshtoken-';
exports.redisUserPrefix = 'skuser-';


exports.skTypes = {
  'totpExternal': 'totpExternal',
  'totp':'totp',
  'u2fHwToken':'u2fHwToken',
  'webAuthNHwToken':'webAuthNHwToken',
  'fidoMobileApp':'fidoMobileApp',
  'sms':'sms',
  symcTotp:'symcTotp'
}




// https://mds.fidoalliance.org/Root.cer
exports.mdsRootCert =
  "-----BEGIN CERTIFICATE-----\n" +
  "MIICQzCCAcigAwIBAgIORqmxkzowRM99NQZJurcwCgYIKoZIzj0EAwMwUzELMAkG\n" +
  "A1UEBhMCVVMxFjAUBgNVBAoTDUZJRE8gQWxsaWFuY2UxHTAbBgNVBAsTFE1ldGFk\n" +
  "YXRhIFRPQyBTaWduaW5nMQ0wCwYDVQQDEwRSb290MB4XDTE1MDYxNzAwMDAwMFoX\n" +
  "DTQ1MDYxNzAwMDAwMFowUzELMAkGA1UEBhMCVVMxFjAUBgNVBAoTDUZJRE8gQWxs\n" +
  "aWFuY2UxHTAbBgNVBAsTFE1ldGFkYXRhIFRPQyBTaWduaW5nMQ0wCwYDVQQDEwRS\n" +
  "b290MHYwEAYHKoZIzj0CAQYFK4EEACIDYgAEFEoo+6jdxg6oUuOloqPjK/nVGyY+\n" +
  "AXCFz1i5JR4OPeFJs+my143ai0p34EX4R1Xxm9xGi9n8F+RxLjLNPHtlkB3X4ims\n" +
  "rfIx7QcEImx1cMTgu5zUiwxLX1ookVhIRSoso2MwYTAOBgNVHQ8BAf8EBAMCAQYw\n" +
  "DwYDVR0TAQH/BAUwAwEB/zAdBgNVHQ4EFgQU0qUfC6f2YshA1Ni9udeO0VS7vEYw\n" +
  "HwYDVR0jBBgwFoAU0qUfC6f2YshA1Ni9udeO0VS7vEYwCgYIKoZIzj0EAwMDaQAw\n" +
  "ZgIxAKulGbSFkDSZusGjbNkAhAkqTkLWo3GrN5nRBNNk2Q4BlG+AvM5q9wa5WciW\n" +
  "DcMdeQIxAMOEzOFsxX9Bo0h4LOFE5y5H8bdPFYW+l5gy1tQiJv+5NUyM2IBB55XU\n" +
  "YjdBz56jSA==\n" +
  "-----END CERTIFICATE-----\n";

// https://mds.fidoalliance.org/Root.crl
exports.mdsRootCrl =
  "-----BEGIN X509 CRL-----\n" +
  "MIIBLTCBswIBATAKBggqhkjOPQQDAzBTMQswCQYDVQQGEwJVUzEWMBQGA1UEChMN\n" +
  "RklETyBBbGxpYW5jZTEdMBsGA1UECxMUTWV0YWRhdGEgVE9DIFNpZ25pbmcxDTAL\n" +
  "BgNVBAMTBFJvb3QXDTE4MDQwNzAwMDAwMFoXDTE4MDcxNTAwMDAwMFqgLzAtMAoG\n" +
  "A1UdFAQDAgEMMB8GA1UdIwQYMBaAFNKlHwun9mLIQNTYvbnXjtFUu7xGMAoGCCqG\n" +
  "SM49BAMDA2kAMGYCMQCnXSfNppE9vpsGtY9DsPWyR3aVVSPs6i5/3A21a1+rCNoa\n" +
  "1cJNWKZJ7IV4cdjIXVUCMQCDh8U8OekdTnuvcG3FaoMJO0y0C0FS5dbTzcuiADjy\n" +
  "VbAQeaSsCauVySzyB3lVVgE=\n" +
  "-----END X509 CRL-----\n";

// http://mds.fidoalliance.org/CA-1.crl
exports.ca1Crl =
  "-----BEGIN X509 CRL-----\n" +
  "MIIBDTCBswIBATAKBggqhkjOPQQDAjBTMQswCQYDVQQGEwJVUzEWMBQGA1UEChMN\n" +
  "RklETyBBbGxpYW5jZTEdMBsGA1UECxMUTWV0YWRhdGEgVE9DIFNpZ25pbmcxDTAL\n" +
  "BgNVBAMTBENBLTEXDTE4MDYwNzAwMDAwMFoXDTE4MDcxNTAwMDAwMFqgLzAtMAoG\n" +
  "A1UdFAQDAgEkMB8GA1UdIwQYMBaAFGkRXi1pZIWdlrjW/1zNvzx1z0wYMAoGCCqG\n" +
  "SM49BAMCA0kAMEYCIQDEDFIsNHgOZUUolm0XIyyGO5Qrr7byVtjfkd7nTfpAlAIh\n" +
  "AIBctNT3uR9vLosOHQexvhp2EL/KO9cALAk6HaVwL/LD\n" +
  "-----END X509 CRL-----\n";

exports.mdsIntermediateCert =
  "-----BEGIN CERTIFICATE-----\n" +
  "MIICsjCCAjigAwIBAgIORqmxk8NQuJfCENVYa1QwCgYIKoZIzj0EAwMwUzELMAkG\n" +
  "A1UEBhMCVVMxFjAUBgNVBAoTDUZJRE8gQWxsaWFuY2UxHTAbBgNVBAsTFE1ldGFk\n" +
  "YXRhIFRPQyBTaWduaW5nMQ0wCwYDVQQDEwRSb290MB4XDTE1MDYxNzAwMDAwMFoX\n" +
  "DTQwMDYxNzAwMDAwMFowUzELMAkGA1UEBhMCVVMxFjAUBgNVBAoTDUZJRE8gQWxs\n" +
  "aWFuY2UxHTAbBgNVBAsTFE1ldGFkYXRhIFRPQyBTaWduaW5nMQ0wCwYDVQQDEwRD\n" +
  "QS0xMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE9sDgC8PzBYl/wKqpXfa98jOI\n" +
  "o78l9pz4xOzGDGIz0zEXMXsBY6kAhyU4GRmT0wo4tyUvx5BY8OKlsLMzlbKMRaOB\n" +
  "7zCB7DAOBgNVHQ8BAf8EBAMCAQYwEgYDVR0TAQH/BAgwBgEB/wIBADAdBgNVHQ4E\n" +
  "FgQUaRFeLWlkhZ2WuNb/XM2/PHXPTBgwHwYDVR0jBBgwFoAU0qUfC6f2YshA1Ni9\n" +
  "udeO0VS7vEYwNQYDVR0fBC4wLDAqoCigJoYkaHR0cDovL21kcy5maWRvYWxsaWFu\n" +
  "Y2Uub3JnL1Jvb3QuY3JsME8GA1UdIARIMEYwRAYLKwYBBAGC5RwBAwEwNTAzBggr\n" +
  "BgEFBQcCARYnaHR0cHM6Ly9tZHMuZmlkb2FsbGlhbmNlLm9yZy9yZXBvc2l0b3J5\n" +
  "MAoGCCqGSM49BAMDA2gAMGUCMBLVq0JdWv2yY4Rp1IiyIVWEKG1PTz1pPAFqEnak\n" +
  "Ptw4RMRTGwHdb2ifcDbPoEkfYQIxAOLkfEPj22fBnej1wtgyylsu73rKLUv4xhDy\n" +
  "9TAeVUml0iDBM8StE4DiVs/4ejFhqQ==\n" +
  "-----END CERTIFICATE-----\n";

exports.mdsSigningCert =
  "-----BEGIN CERTIFICATE-----\n" +
  "MIICnTCCAkOgAwIBAgIORvCM1auU6FYVXUebJHcwCgYIKoZIzj0EAwIwUzELMAkG\n" +
  "A1UEBhMCVVMxFjAUBgNVBAoTDUZJRE8gQWxsaWFuY2UxHTAbBgNVBAsTFE1ldGFk\n" +
  "YXRhIFRPQyBTaWduaW5nMQ0wCwYDVQQDEwRDQS0xMB4XDTE1MDgxOTAwMDAwMFoX\n" +
  "DTE4MDgxOTAwMDAwMFowZDELMAkGA1UEBhMCVVMxFjAUBgNVBAoTDUZJRE8gQWxs\n" +
  "aWFuY2UxHTAbBgNVBAsTFE1ldGFkYXRhIFRPQyBTaWduaW5nMR4wHAYDVQQDExVN\n" +
  "ZXRhZGF0YSBUT0MgU2lnbmVyIDMwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAASK\n" +
  "X+p3W2j1GV4lQwn7HXNj4lh9e2wAa6J9tBIQhbQTkqMvNZGnHxOn7yTZ3NpYO5ZG\n" +
  "Vgr/XC66qli7BWA8jgTfo4HpMIHmMA4GA1UdDwEB/wQEAwIGwDAMBgNVHRMBAf8E\n" +
  "AjAAMB0GA1UdDgQWBBRckNF+zzxMuLvm+qRjLeJQf0DwyzAfBgNVHSMEGDAWgBRp\n" +
  "EV4taWSFnZa41v9czb88dc9MGDA1BgNVHR8ELjAsMCqgKKAmhiRodHRwOi8vbWRz\n" +
  "LmZpZG9hbGxpYW5jZS5vcmcvQ0EtMS5jcmwwTwYDVR0gBEgwRjBEBgsrBgEEAYLl\n" +
  "HAEDATA1MDMGCCsGAQUFBwIBFidodHRwczovL21kcy5maWRvYWxsaWFuY2Uub3Jn\n" +
  "L3JlcG9zaXRvcnkwCgYIKoZIzj0EAwIDSAAwRQIhALLbYjBrbhPkwrn3mQjCERIw\n" +
  "kMNNT/lfkpNXH+4zjUXEAiBas2lP6jp44Bh4X+tBXqY7y61ijGRIZCaAF1KIlgub\n" +
  "0g==\n" +
  "-----END CERTIFICATE-----\n";

exports.HASHALG = {
  [-7]: { algName:'ECDSA_w_SHA-256' ,hashAlg:'SHA256' },
  [-35]: { algName:'ECDSA_w_SHA384' ,hashAlg:'SHA384' },
  [-36] : { algName: "ECDSA_w_SHA512", hashAlg: "SHA512"},

  [-8]: { algName:'EdDSA signature algorithms' ,hashAlg:'SHA256' },

  [-65535] : { algName: "RSA_w_SHA1", hashAlg: "SHA1"},
  [-257] : { algName: "RSA_w_SHA256", hashAlg: "SHA256"},
  [-258] : { algName: "RSA_w_SHA384", hashAlg: "SHA384"},
  [-259] : { algName: "RSA_w_SHA512", hashAlg: "SHA512"},

  [-37] : { algName: "RSASSA-PSS w/ SHA-256", hashAlg: "SHA256"},
  [-38] : { algName: "RSASSA-PSS w/ SHA-384", hashAlg: "SHA384"},
  [-39] : { algName: "RSASSA-PSS w/ SHA-512", hashAlg: "SHA512"},

}



exports.FLAGS = {
  U2F_USER_PRESENTED: 0x01,
  RFU1: 0x02,
  UV: 0x04,
  RFU3: 0x08,
  RFU4: 0x10,
  RFU5: 0x20,
  AT: 0x40,
  ED: 0x80,
}

exports.PROCESS_EXIT_BLACKLIST = ['oe'];
